instance/models/instance.py

```diff
@@ -455,29 +461,53 @@ def run_playbook(self):
         """
         Run a playbook against the instance active servers
         """
-        with open_repository(self.ansible_source_repo_url,
-                             ref=self.configuration_version) as configuration_repo:
-            playbook_path = os.path.join(configuration_repo.working_dir, 'playbooks')
-            requirements_path = os.path.join(configuration_repo.working_dir, 'requirements.txt')
-
-            self.logger.info('Running playbook "%s/%s"...', playbook_path, self.ansible_playbook_filename)
-
-            log_lines = []
-            with ansible.run_playbook(
-                requirements_path,
-                self.inventory_str,
-                self.vars_str,
-                playbook_path,
-                self.ansible_playbook_filename,
-                username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
-            ) as processus:
-                for line in processus.stdout:
-                    line = line.decode('utf-8').rstrip()
-                    self.logger.info(line)
-                    log_lines.append([line.rstrip()])
-
-        self.logger.info('Playbook run completed')
-        return log_lines
+        attempt = 0
+        while attempt < self.attempts:
+            with open_repository(self.ansible_source_repo_url,
+                                 ref=self.configuration_version) as configuration_repo:
+                playbook_path = os.path.join(configuration_repo.working_dir,
+                                             'playbooks')
+                requirements_path = os.path.join(configuration_repo.working_dir,
+                                                 'requirements.txt')
+
+                log = ('Running playbook "{path}/{name}" for instance {instance}, '
+                       'attempt {attempt} of {attempts}:').format(
+                           path=playbook_path,
+                           name=self.ansible_playbook_name,
+                           instance=self,
+                           attempts=self.attempts,
+                           attempt=attempt + 1)
+                self.logger.info(log)
+
+                log_lines = []
+                with ansible.run_playbook(
+                    requirements_path,
+                    self.inventory_str,
+                    self.vars_str,
+                    playbook_path,
+                    self.ansible_playbook_filename,
+                    username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
+                ) as processus:
+                    output = itertools.chain(
+                        zip(itertools.repeat(logging.INFO), processus.stdout),
+                        zip(itertools.repeat(logging.ERROR), processus.stderr)
+                    )
```

Comment:
@chromano This is a very generic implementation for reading from multiple file descriptors, and it would certainly work for this use case.  If we just yield `line`, we won't be able to distinguish between the different streams, though, which is the only point of this approach (see my comment on using `stderr=subprocess.STDOUT` above).  So we need to include the information which stream the line came from to set the log level accordingly.

I just noticed there is a [`selectors`](https://docs.python.org/3/library/selectors.html) module in Python 3.4 I wasn't aware of before.  You can use `selectors.DefaultSelector()` instead of `select.poll()` for better portability – the interface is quite similar.

There definitely are truly asynchronous approaches, by the way.  Using threads is the most obvious one, but since Python 3.4 there is also the event-loop based `asyncio` module – see https://docs.python.org/3.4/library/asyncio-subprocess.html for detail on how to use this with subprocesses.  (I'm not suggesting to use it here – `poll()` should work fine.)